src/common/hist_util.cc

```diff
@@ -248,17 +247,20 @@ void GHistIndexMatrix::Init(DMatrix* p_fmat, int max_num_bins) {
   size_t prev_sum = 0;
 
   for (const auto &batch : p_fmat->GetRowBatches()) {
+    auto bsize = static_cast<omp_ulong>(batch.Size());
+    nthread = std::min(bsize, nthread);
```

Comment:
Wouldn't this make `nthread` too small if batch size fluctuates? Probably should declare a new variable.

Reply:
i think this is going to be pertinent only for the last batch. for the ones before that, nthread will be the default omp max. for the last batch, *if* it happens to be lesser (typically ~ 64), then it scales it down. the nthread is mostly pegged to the batch size in this block.

Reply:
From tracing through some tests, the batch size doesn't seem to be constant and can fluctuate up and down. Seems safer to not make nthread a running min.

Reply:
i have left the number of threads as is and adjusted the partial sums if the batch size is smaller than the number of threads.

the test should also now test all possible scenarios with batch size and thread pool size.